File: DSA/binary_trees/treeSum.js
class Node {
	constructor(val) {
		this.val = val
		this.left = null
		this.right = null
	}
}

const a = new Node(3)
const b = new Node(11)
const c = new Node(4)
const d = new Node(4)
const e = new Node(2)
const f = new Node(1)

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

const treeSum = (root) => {
	if (root === null) return 0
	let sum = 0
	const queue = [root]

	while (queue.length > 0) {
		const current = queue.shift()
		sum += current.val
		if (current.left) queue.push(current.left)
		if (current.right) queue.push(current.right)
	}
	return sum
}

const treeSumRec = (root) => {
	if (root == null) return 0
	return root.val + treeSum(root.left) + treeSum(root.right)
}

console.log('Getting sum iteratively')
console.log(treeSum(a))
console.log('Getting sum recursively')
console.log(treeSumRec(a))

File: DSA/data_structures/binary_trees/tree_sum.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

a = Node(3)
b = Node(11)
c = Node(4)
d = Node(4)
e = Node(2)
f = Node(1)

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

def tree_sum(root):
    if root == None: return 0
    sum = 0
    queue = [root]

    while queue:
        current = queue.pop(0)
        sum += current.val
        if current.left: queue.append(current.left)
        if current.right: queue.append(current.right)
    return sum

def tree_sum_rec(root):
    if root == None: return 0
    return root.val + tree_sum(root.left) + tree_sum(root.right)

print('Getting sum iteratively')
print(tree_sum(a))
print('Getting sum recursively')
print(tree_sum_rec(a))

